### PR TITLE
feat(schema): pure cost models for a benchmark run (ENG-61)

### PR DESCRIPTION
**Pull request by dbworku**
**ENG-61 economics — split into 3 focused PRs** (merge bottom-up, each independently green):
1. **#80 — pure cost models** (this PR)
2. #71 — economics Catalog slice + derivation
3. #81 — derive economics onto each Run

↑ **This PR is 1/3**, based on #68.

---
The **pure pricing math** the economics dimension rests on — burst (pay-per-use) and fixed-infra amortized cost models, plus their break-even utilization. Zero schema dependencies; unit-tested in isolation. The MetricDefs + `deriveEconomics()` that consume `burstCostPerRun` land in #71.

**Validated locally:** `bun run typecheck` (8 workspaces, 0 errors); `@sandbox-benchmarks/schema` 122 tests pass.

---
## Summary by cubic
Adds pure per-run cost models for the economics dimension: burst (pay-per-use), fixed-infra amortization, and a break-even runs/month helper. Exported from `index.ts`; supports ENG-61.

- **New Features**
  - Helpers: `burstCostPerRun(hourlyUsd, runtimeMs)`, `amortizedCostPerRun(monthlyUsd, runsPerMonth)` (`Infinity` when runs/month ≤ 0), `amortizationBreakEvenRunsPerMonth(monthlyUsd, hourlyUsd, runtimeMs)` (`Infinity` when burst cost is 0).
  - Unit tests in `economics.test.ts`.

<sup>Written for commit 6f5a6a3b7ba58d58bb319289b01f6211a57f9f29. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/starslingdev/sandbox-benchmarks/pull/80?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added economics cost utilities to estimate burst per-run costs, amortized per-run costs (with `Infinity` for zero/negative usage), and the monthly runs-per-month break-even point.
  * Re-exported these new pricing helpers as part of the public schema API.
* **Tests**
  * Added Bun-based unit tests covering burst, amortized, and break-even calculations, including edge cases like zero runtime and zero usage.

### CODE REVIEW

**Comment by dbworku** [comment]
* **#73** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/73?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#74](https://github.com/starslingdev/sandbox-benchmarks/pull/74) <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/74?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#75](https://github.com/starslingdev/sandbox-benchmarks/pull/75) <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/75?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#72** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/72?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#83** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/83?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#82** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/82?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#81** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/81?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#71** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/71?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#80** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/80?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/80?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#68** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/68?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#84](https://github.com/starslingdev/sandbox-benchmarks/pull/84) <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/84?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#79** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/79?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#67** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/67?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#78** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/78?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#88** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/88?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>merge-ready</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/starslingdev?ref=merge-queue-instructions-in-stack-comment&prId=80).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/schema/src/economics.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+const MS_PER_HOUR = 3_600_000;
+
+/**
+ * Burst (pay-per-use) cost of one run: you pay only for the wall-clock the sandbox is alive, so the
+ * cost scales linearly with runtime. The model behind every runtime economics Metric
+ * (`usd_per_lifecycle`, `usd_per_compute_run`) — the serverless/per-second default.
+ */
+export function burstCostPerRun(hourlyUsd: number, runtimeMs: number): number {
+	return hourlyUsd * (runtimeMs / MS_PER_HOUR);
+}
```
[the remark this answers is not in the document]
Comment (dbworku) [reply packages/schema/src/economics.ts:16]: By construction: burst inputs are non-negative (measured runtimeMs, a vetted positive hourly rate). A negative would signal an upstream bug better surfaced than silently clamped, so the pure helper stays unguarded.

File: packages/schema/src/economics.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * the cost *attributable* to a single run is that fixed bill spread across the runs it serves that
+ * month. Per-run cost therefore FALLS as utilization rises — the mirror image of
+ * {@link burstCostPerRun}. `runsPerMonth <= 0` returns `Infinity` (a box that serves no runs
+ * amortizes its bill over nothing). Ported in spirit from THEIRS `compute_costs` `ec2_full`
+ * (variable per-run + a `monthly_fixed` overhead), kept as a documented helper rather than a
+ * catalogued Metric since OURS prices no reserved infra yet.
+ */
+export function amortizedCostPerRun(monthlyUsd: number, runsPerMonth: number): number {
+	return runsPerMonth > 0 ? monthlyUsd / runsPerMonth : Number.POSITIVE_INFINITY;
+}
```
[the remark this answers is not in the document]
Comment (dbworku) [reply packages/schema/src/economics.ts:29]: Same rationale: monthlyUsd is a vetted non-negative rate; amortizedCostPerRun already guards the meaningful edge (runsPerMonth<=0 → Infinity). Leaving the non-negativity to the caller's inputs.

File: packages/schema/src/economics.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * Returns `Infinity` when a run's burst cost is 0 (no per-run cost to amortize against).
+ */
+export function amortizationBreakEvenRunsPerMonth(
+	monthlyUsd: number,
+	hourlyUsd: number,
+	runtimeMs: number,
+): number {
+	const burst = burstCostPerRun(hourlyUsd, runtimeMs);
+	return burst > 0 ? monthlyUsd / burst : Number.POSITIVE_INFINITY;
+}
```
[the remark this answers is not in the document]
Comment (dbworku) [reply packages/schema/src/economics.ts:45]: Same as the other cost-model helpers — inputs are non-negative by construction; not adding input validation to the pure math.

File: packages/schema/src/economics.test.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	burstCostPerRun,
+} from "./index.ts";
+
+describe("cost-model helpers (burst vs fixed-infra amortization)", () => {
+	it("burstCostPerRun scales linearly with runtime", () => {
+		// $3.60/hr → $0.001/s; a 10s run costs $0.01, double the runtime → double the cost.
+		expect(burstCostPerRun(3.6, 10_000)).toBeCloseTo(0.01, 12);
+		expect(burstCostPerRun(3.6, 20_000)).toBeCloseTo(0.02, 12);
+		expect(burstCostPerRun(3.6, 0)).toBe(0);
+	});
```
[the remark this answers is not in the document]
Comment (dbworku) [reply packages/schema/src/economics.test.ts:14]: Won't add a negative-runtime test: unlike the runsPerMonth<=0 divide-by-zero (a real reachable edge), a negative runtimeMs is not a value the normalizer can produce, so there's no behavior worth pinning.

File: packages/schema/src/index.ts [hunk 5]
```diff
@@ -7,6 +7,8 @@ import { rawRunSchema } from "./lib/internal.ts";
 export * from "./analysis.ts";
 // The Metric Catalog — the registry of rankable Metrics, plus lookup helpers.
 export * from "./catalog.ts";
+// The derived economics Dimension ($/run): its MetricDefs and the pricing-driven derivation.
+export * from "./economics.ts";
```
[the remark this answers is not in the document]
Comment (dbworku) [reply packages/schema/src/index.ts]: Fixed: the barrel comment now describes the cost models on this PR; it's expanded to MetricDefs + derivation on #71 where those land.

**Review by cosmic-emerald — APPROVED** [review]
